Add a subcommand to display a hexdump of transceiver EEPROM page

mihirpat1 opened the pull request:
**What I did**
Added a subcommand to display the hexdump of transceiver EEPROM page specified by the user.
"sfputil show eeprom-hexdump -p <port_name> -n <page_number>" has been added for this purpose. Specifying the port name is mandatory. If page number is not mentioned, page 0 content will be displayed by default.
For SFF-8472 based transceivers, following is the behavior
The entire 256 byte region of A0h will displayed. In addition to this, the lower 128 byte of A2h will be displayed.
The upper 128 bytes of A2h will be displayed based on the page number specified by the user. If no page number is specified, page 0 relevant data will be shown.

For non SFF-8472 based transceiver, following is the behavior
Lower page 0h is always displayed and the upper page is displayed based on the page number specified by the user

**How to verify it**
Verified the utility with 400G DD, 100G QSFP28, 40G DAC and 10G transceiver
```
400G DD
Page 0
CLI
sfputil show eeprom-hexdump -p Ethernet16 -n 0
root@str2-z9332f-05:/usr/local/lib/python3.9/dist-packages/sfputil# sfputil show eeprom-hexdump -p Ethernet16 -n 0
EEPROM hexdump for port Ethernet16 page 0h
        Lower page 0h
        00000000 18 30 80 00 00 00 00  00 00 00 00 00 00 00 00 00 |.0..............|
        00000010 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000020 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000030 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000040 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000050 00 00 00 00 00 03 1d  01 88 01 ff 00 00 00 00 00 |................|
        00000060 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000070 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|

        Upper page 0h
        00000080 18 44 45 4c 4c 20 45  4d 43 20 20 20 20 20 20 20 |.DELL EMC       |
        00000090 20 3c 18 a0 32 32 56  36 52 20 20 20 20 20 20 20 | <..22V6R       |
        000000a0 20 20 20 20 41 30 43  4e 30 4c 58 44 30 30 30 36 |    A0CN0LXD0006|
        000000b0 4b 49 39 35 33 20 32  30 30 36 32 30 20 20 00 00 |KI953 200620  ..|
        000000c0 00 00 00 00 00 00 00  00 00 02 41 23 05 06 0a 16 |..........A#....|
        000000d0 00 00 00 02 0a 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000e0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000f0 df 10 04 93 01 00 00  00 00 00 00 00 00 00 00 00 |................|

root@str2-z9332f-05:/usr/local/lib/python3.9/dist-packages/sfputil# 

File dump
hexdump -C -n 256 12-0050/eeprom 
root@str2-z9332f-05:/sys/bus/i2c/devices# hexdump -C -n 256 12-0050/eeprom 
00000000  18 30 80 00 00 00 00 00  00 00 00 00 00 00 00 00  |.0..............|
00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000050  00 00 00 00 00 03 1d 01  88 01 ff 00 00 00 00 00  |................|
00000060  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000080  18 44 45 4c 4c 20 45 4d  43 20 20 20 20 20 20 20  |.DELL EMC       |
00000090  20 3c 18 a0 32 32 56 36  52 20 20 20 20 20 20 20  | <..22V6R       |
000000a0  20 20 20 20 41 30 43 4e  30 4c 58 44 30 30 30 36  |    A0CN0LXD0006|
000000b0  4b 49 39 35 33 20 32 30  30 36 32 30 20 20 00 00  |KI953 200620  ..|
000000c0  00 00 00 00 00 00 00 00  00 02 41 23 05 06 0a 16  |..........A#....|
000000d0  00 00 00 02 0a 00 00 00  00 00 00 00 00 00 00 00  |................|
000000e0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000f0  df 10 04 93 01 00 00 00  00 00 00 00 00 00 00 00  |................|
00000100
root@str2-z9332f-05:/sys/bus/i2c/devices# 

Page 2
CLI
sfputil show eeprom-hexdump -p Ethernet16 -n 2
root@str2-z9332f-05:/usr/local/lib/python3.9/dist-packages/sfputil# sfputil show eeprom-hexdump -p Ethernet16 -n 2
EEPROM hexdump for port Ethernet16 page 2h
        Lower page 0h
        00000000 18 30 80 00 00 00 00  00 00 00 00 00 00 00 00 00 |.0..............|
        00000010 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000020 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000030 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000040 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000050 00 00 00 00 00 03 1d  01 88 01 ff 00 00 00 00 00 |................|
        00000060 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000070 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|

        Upper page 2h

root@str2-z9332f-05:/usr/local/lib/python3.9/dist-packages/sfputil# 

File dump
hexdump -C -n 512 12-0050/eeprom 
root@str2-z9332f-05:/sys/bus/i2c/devices# hexdump -C -n 512 12-0050/eeprom 
00000000  18 30 80 00 00 00 00 00  00 00 00 00 00 00 00 00  |.0..............|
00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000050  00 00 00 00 00 03 1d 01  88 01 ff 00 00 00 00 00  |................|
00000060  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000080  18 44 45 4c 4c 20 45 4d  43 20 20 20 20 20 20 20  |.DELL EMC       |
00000090  20 3c 18 a0 32 32 56 36  52 20 20 20 20 20 20 20  | <..22V6R       |
000000a0  20 20 20 20 41 30 43 4e  30 4c 58 44 30 30 30 36  |    A0CN0LXD0006|
000000b0  4b 49 39 35 33 20 32 30  30 36 32 30 20 20 00 00  |KI953 200620  ..|
000000c0  00 00 00 00 00 00 00 00  00 02 41 23 05 06 0a 16  |..........A#....|
000000d0  00 00 00 02 0a 00 00 00  00 00 00 00 00 00 00 00  |................|
000000e0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000f0  df 10 04 93 01 00 00 00  00 00 00 00 00 00 00 00  |................|
00000100
root@str2-z9332f-05:/sys/bus/i2c/devices# 

100G QSFP28
str2-a7280cr3-3
Page 0
CLI
sfputil show eeprom-hexdump -p Ethernet0 -n 0
root@str2-a7280cr3-3:/usr/local/lib/python3.9/dist-packages/sfputil# sfputil show eeprom-hexdump -p Ethernet0 -n 0
EEPROM hexdump for port Ethernet0 page 0h
        Lower page 0h
        00000000 11 08 00 10 00 10 00  00 00 00 00 00 00 00 00 00 |................|
        00000010 00 00 00 00 00 00 13  85 00 00 80 d3 00 00 00 00 |................|
        00000020 00 00 32 c9 34 d9 2b  b9 2f 43 00 00 0f 9f 0f 80 |..2.4.+./C......|
        00000030 0f 80 00 01 36 17 34  d2 30 43 00 00 00 00 00 00 |....6.4.0C......|
        00000040 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000050 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000060 00 00 ff 00 00 00 00  00 00 00 00 00 00 04 08 00 |................|
        00000070 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|

        Upper page 0h
        00000080 11 8c 23 80 00 00 00  00 00 00 00 07 ff 00 00 00 |..#.............|
        00000090 00 00 05 00 46 49 4e  49 53 41 52 20 43 4f 52 50 |....FINISAR CORP|
        000000a0 20 20 20 20 00 00 90  65 46 43 42 52 34 32 35 51 |    ...eFCBR425Q|
        000000b0 45 32 43 30 35 2d 4d  53 41 30 42 68 00 00 00 30 |E2C05-MSA0Bh...0|
        000000c0 01 07 ff dc 55 33 53  41 48 36 37 20 20 20 20 20 |....U3SAH67     |
        000000d0 20 20 20 20 32 30 30  37 30 37 20 20 38 00 67 e3 |    200707  8.g.|
        000000e0 20 20 20 20 20 20 20  20 20 20 20 20 20 20 20 20 |                |
        000000f0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|

root@str2-a7280cr3-3:/usr/local/lib/python3.9/dist-packages/sfputil# 

File dump
hexdump -C -n 256 15-0050/eeprom
root@str2-a7280cr3-3:/sys/bus/i2c/devices# hexdump -C -n 256 15-0050/eeprom
00000000  11 08 00 10 00 10 00 00  00 00 00 00 00 00 00 00  |................|
00000010  00 00 00 00 00 00 13 c3  00 00 80 f4 00 00 00 00  |................|
00000020  00 00 32 00 34 ed 34 b2  30 3e 00 00 0f 9f 0f 80  |..2.4.4.0>......|
00000030  0f 80 00 01 36 14 34 ce  30 41 00 00 00 00 00 00  |....6.4.0A......|
00000040  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000060  00 00 ff 00 00 00 00 00  00 00 00 00 00 04 08 00  |................|
00000070  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000080  11 8c 23 80 00 00 00 00  00 00 00 07 ff 00 00 00  |..#.............|
00000090  00 00 05 00 46 49 4e 49  53 41 52 20 43 4f 52 50  |....FINISAR CORP|
000000a0  20 20 20 20 00 00 90 65  46 43 42 52 34 32 35 51  |    ...eFCBR425Q|
000000b0  45 32 43 30 35 2d 4d 53  41 30 42 68 00 00 00 30  |E2C05-MSA0Bh...0|
000000c0  01 07 ff dc 55 33 53 41  48 36 37 20 20 20 20 20  |....U3SAH67     |
000000d0  20 20 20 20 32 30 30 37  30 37 20 20 38 00 67 e3  |    200707  8.g.|
000000e0  20 20 20 20 20 20 20 20  20 20 20 20 20 20 20 20  |                |
000000f0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000100
root@str2-a7280cr3-3:/sys/bus/i2c/devices# 

Page 3
CLI
sfputil show eeprom-hexdump -p Ethernet0 -n 3
root@str2-a7280cr3-3:/sys/bus/i2c/devices# sfputil show eeprom-hexdump -p Ethernet0 -n 3
EEPROM hexdump for port Ethernet0 page 3h
        Lower page 0h
        00000000 11 08 00 10 00 10 00  00 00 00 00 00 00 00 00 00 |................|
        00000010 00 00 00 00 00 00 13  ab 00 00 81 12 00 00 00 00 |................|
        00000020 00 00 32 c2 35 55 35  80 2f b5 00 00 0f 9f 0f 80 |..2.5U5./.......|
        00000030 0f 80 00 01 36 15 34  cf 30 42 00 00 00 00 00 00 |....6.4.0B......|
        00000040 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000050 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000060 00 00 ff 00 00 00 00  00 00 00 00 00 00 04 08 00 |................|
        00000070 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|

        Upper page 3h
        00000080 4b 00 fb 00 46 00 00  00 00 00 00 00 00 00 00 00 |K...F...........|
        00000090 8d cc 74 04 87 5a 79  18 00 00 00 00 00 00 00 00 |..t..Zy.........|
        000000a0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000b0 55 76 01 8e 43 e2 03  1a 17 70 05 dc 15 7c 07 d0 |Uv..C....p...|..|
        000000c0 7b 87 03 e8 4d f1 06  31 00 00 00 00 00 00 00 00 |{...M..1........|
        000000d0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000e0 a7 0f 00 00 00 00 00  00 00 00 22 22 22 22 33 33 |..........""""33|
        000000f0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|

root@str2-a7280cr3-3:/sys/bus/i2c/devices# 

File dump
hexdump -C -n 640 15-0050/eeprom
root@str2-a7280cr3-3:/sys/bus/i2c/devices# hexdump -C -n 640 15-0050/eeprom
00000000  11 08 00 10 00 10 00 00  00 00 00 00 00 00 00 00  |................|
00000010  00 00 00 00 00 00 13 cb  00 00 80 83 00 00 00 00  |................|
00000020  00 00 32 4e 35 14 34 df  2f c7 00 00 0f 9f 0f 80  |..2N5.4./.......|
00000030  0f 80 00 01 36 14 34 ce  30 41 00 00 00 00 00 00  |....6.4.0A......|
00000040  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000060  00 00 ff 00 00 00 00 00  00 00 00 00 00 04 08 00  |................|
00000070  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000080  11 8c 23 80 00 00 00 00  00 00 00 07 ff 00 00 00  |..#.............|
00000090  00 00 05 00 46 49 4e 49  53 41 52 20 43 4f 52 50  |....FINISAR CORP|
000000a0  20 20 20 20 00 00 90 65  46 43 42 52 34 32 35 51  |    ...eFCBR425Q|
000000b0  45 32 43 30 35 2d 4d 53  41 30 42 68 00 00 00 30  |E2C05-MSA0Bh...0|
000000c0  01 07 ff dc 55 33 53 41  48 36 37 20 20 20 20 20  |....U3SAH67     |
000000d0  20 20 20 20 32 30 30 37  30 37 20 20 38 00 67 e3  |    200707  8.g.|
000000e0  20 20 20 20 20 20 20 20  20 20 20 20 20 20 20 20  |                |
000000f0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000200  4b 00 fb 00 46 00 00 00  00 00 00 00 00 00 00 00  |K...F...........|
00000210  8d cc 74 04 87 5a 79 18  00 00 00 00 00 00 00 00  |..t..Zy.........|
00000220  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000230  55 76 01 8e 43 e2 03 1a  17 70 05 dc 15 7c 07 d0  |Uv..C....p...|..|
00000240  7b 87 03 e8 4d f1 06 31  00 00 00 00 00 00 00 00  |{...M..1........|
00000250  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000260  a7 0f 00 00 00 00 00 00  00 00 22 22 22 22 33 33  |..........""""33|
00000270  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000280
root@str2-a7280cr3-3:/sys/bus/i2c/devices# 

10G
str-7260cx3-acs-7
sfputil show eeprom-hexdump -p Ethernet256 -n 0
root@str-7260cx3-acs-7:/usr/local/lib/python3.9/dist-packages/sfputil# sfputil show eeprom-hexdump -p Ethernet256 -n 0
EEPROM hexdump for port Ethernet256 page 0h
        A0h dump
        00000000 03 04 07 10 00 00 00  00 00 00 00 06 67 00 00 00 |............g...|
        00000010 08 03 00 1e 46 49 4e  49 53 41 52 20 43 4f 52 50 |....FINISAR CORP|
        00000020 2e 20 20 20 00 00 90  65 46 54 4c 58 38 35 37 31 |.   ...eFTLX8571|
        00000030 44 33 42 43 4c 20 20  20 41 20 20 20 03 52 00 48 |D3BCL   A   .R.H|
        00000040 00 1a 00 00 41 55 4a  30 52 43 4a 20 20 20 20 20 |....AUJ0RCJ     |
        00000050 20 20 20 20 31 35 31  30 32 39 20 20 68 f0 03 f6 |    151029  h...|
        00000060 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000070 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000080 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000090 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000a0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000b0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000c0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000d0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000e0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000f0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|

        A2h dump (lower 128 bytes)
        00000000 4e 00 f3 00 49 00 f8  00 90 88 71 48 8c a0 75 30 |N...I.....qH..u0|
        00000010 19 c8 07 d0 18 9c 09  c4 27 10 09 d0 1f 07 0c 5a |........'......Z|
        00000020 27 10 00 64 1f 07 00  9e 00 00 00 00 00 00 00 00 |'..d............|
        00000030 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000040 00 00 00 00 3f 80 00  00 00 00 00 00 01 00 00 00 |....?...........|
        00000050 01 00 00 00 01 00 00  00 01 00 00 00 00 00 00 1b |................|
        00000060 14 00 81 ad 0d 5f 17  a4 15 18 00 00 00 00 30 00 |....._........0.|
        00000070 00 00 00 00 00 00 00  00 ff ff ff ff ff ff ff 01 |................|

        A2h dump (upper 128 bytes) page 0h
        00000080 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000090 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000a0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000b0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000c0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000d0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000e0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000f0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|

root@str-7260cx3-acs-7:/usr/local/lib/python3.9/dist-packages/sfputil# 

File dump
hexdump -C -n 512 7-0050/eeprom
root@str-7260cx3-acs-7:/sys/bus/i2c/devices# hexdump -C -n 512 7-0050/eeprom
00000000  03 04 07 10 00 00 00 00  00 00 00 06 67 00 00 00  |............g...|
00000010  08 03 00 1e 46 49 4e 49  53 41 52 20 43 4f 52 50  |....FINISAR CORP|
00000020  2e 20 20 20 00 00 90 65  46 54 4c 58 38 35 37 31  |.   ...eFTLX8571|
00000030  44 33 42 43 4c 20 20 20  41 20 20 20 03 52 00 48  |D3BCL   A   .R.H|
00000040  00 1a 00 00 41 55 4a 30  52 43 4a 20 20 20 20 20  |....AUJ0RCJ     |
00000050  20 20 20 20 31 35 31 30  32 39 20 20 68 f0 03 f6  |    151029  h...|
00000060  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000100  4e 00 f3 00 49 00 f8 00  90 88 71 48 8c a0 75 30  |N...I.....qH..u0|
00000110  19 c8 07 d0 18 9c 09 c4  27 10 09 d0 1f 07 0c 5a  |........'......Z|
00000120  27 10 00 64 1f 07 00 9e  00 00 00 00 00 00 00 00  |'..d............|
00000130  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000140  00 00 00 00 3f 80 00 00  00 00 00 00 01 00 00 00  |....?...........|
00000150  01 00 00 00 01 00 00 00  01 00 00 00 00 00 00 1b  |................|
00000160  14 73 81 ab 0d 5d 17 7e  15 1a 00 00 00 00 30 00  |.s...].~......0.|
00000170  00 00 00 00 00 00 00 00  ff ff ff ff ff ff ff 01  |................|
00000180  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000200
root@str-7260cx3-acs-7:/sys/bus/i2c/devices# 

Page 2
sfputil show eeprom-hexdump -p Ethernet256 -n 2
root@str-7260cx3-acs-7:/usr/local/lib/python3.9/dist-packages/sfputil# sfputil show eeprom-hexdump -p Ethernet256 -n 2
EEPROM hexdump for port Ethernet256 page 2h
        A0h dump
        00000000 03 04 07 10 00 00 00  00 00 00 00 06 67 00 00 00 |............g...|
        00000010 08 03 00 1e 46 49 4e  49 53 41 52 20 43 4f 52 50 |....FINISAR CORP|
        00000020 2e 20 20 20 00 00 90  65 46 54 4c 58 38 35 37 31 |.   ...eFTLX8571|
        00000030 44 33 42 43 4c 20 20  20 41 20 20 20 03 52 00 48 |D3BCL   A   .R.H|
        00000040 00 1a 00 00 41 55 4a  30 52 43 4a 20 20 20 20 20 |....AUJ0RCJ     |
        00000050 20 20 20 20 31 35 31  30 32 39 20 20 68 f0 03 f6 |    151029  h...|
        00000060 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000070 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000080 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000090 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000a0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000b0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000c0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000d0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000e0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        000000f0 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|

        A2h dump (lower 128 bytes)
        00000000 4e 00 f3 00 49 00 f8  00 90 88 71 48 8c a0 75 30 |N...I.....qH..u0|
        00000010 19 c8 07 d0 18 9c 09  c4 27 10 09 d0 1f 07 0c 5a |........'......Z|
        00000020 27 10 00 64 1f 07 00  9e 00 00 00 00 00 00 00 00 |'..d............|
        00000030 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 00 |................|
        00000040 00 00 00 00 3f 80 00  00 00 00 00 00 01 00 00 00 |....?...........|
        00000050 01 00 00 00 01 00 00  00 01 00 00 00 00 00 00 1b |................|
        00000060 14 80 81 ad 0d 59 17  75 15 11 00 00 00 00 30 00 |.....Y.u......0.|
        00000070 00 00 00 00 00 00 00  00 ff ff ff ff ff ff ff 01 |................|

        A2h dump (upper 128 bytes) page 2h

root@str-7260cx3-acs-7:/usr/local/lib/python3.9/dist-packages/sfputil# 

File dump
hexdump -C 7-0050/eeprom
root@str-7260cx3-acs-7:/sys/bus/i2c/devices# hexdump -C 7-0050/eeprom
00000000  03 04 07 10 00 00 00 00  00 00 00 06 67 00 00 00  |............g...|
00000010  08 03 00 1e 46 49 4e 49  53 41 52 20 43 4f 52 50  |....FINISAR CORP|
00000020  2e 20 20 20 00 00 90 65  46 54 4c 58 38 35 37 31  |.   ...eFTLX8571|
00000030  44 33 42 43 4c 20 20 20  41 20 20 20 03 52 00 48  |D3BCL   A   .R.H|
00000040  00 1a 00 00 41 55 4a 30  52 43 4a 20 20 20 20 20  |....AUJ0RCJ     |
00000050  20 20 20 20 31 35 31 30  32 39 20 20 68 f0 03 f6  |    151029  h...|
00000060  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000100  4e 00 f3 00 49 00 f8 00  90 88 71 48 8c a0 75 30  |N...I.....qH..u0|
00000110  19 c8 07 d0 18 9c 09 c4  27 10 09 d0 1f 07 0c 5a  |........'......Z|
00000120  27 10 00 64 1f 07 00 9e  00 00 00 00 00 00 00 00  |'..d............|
00000130  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000140  00 00 00 00 3f 80 00 00  00 00 00 00 01 00 00 00  |....?...........|
00000150  01 00 00 00 01 00 00 00  01 00 00 00 00 00 00 1b  |................|
00000160  14 27 81 ad 0d 62 17 68  15 0f 00 00 00 00 30 00  |.'...b.h......0.|
00000170  00 00 00 00 00 00 00 00  ff ff ff ff ff ff ff 01  |................|
00000180  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000200
root@str-7260cx3-acs-7:/sys/bus/i2c/devices# 

sfputil show eeprom -p Ethernet256
root@str-7260cx3-acs-7:/usr/local/lib/python3.9/dist-packages/sfputil# sfputil show eeprom -p Ethernet256
Ethernet256: SFP EEPROM detected
        Application Advertisement: N/A
        Connector: LC
        Encoding: 64B/66B
        Extended Identifier: GBIC/SFP defined by two-wire interface ID
        Extended RateSelect Compliance: Unknown
        Identifier: SFP/SFP+/SFP28
        Length OM3(10m): 30.0
        Nominal Bit Rate(100Mbs): 103
        Specification compliance:
                10G Ethernet Compliance: Unknown
                ESCON Compliance: Unknown
                Ethernet Compliance: Unknown
                Fibre Channel Link Length: Unknown
                Fibre Channel Speed: Unknown
                Fibre Channel Transmission Media: Unknown
                Fibre Channel Transmitter Technology: Unknown
                Infiniband Compliance: Unknown
                SFP+CableTechnology: Unknown
                SONET Compliance Codes: Unknown
        Vendor Date Code(YYYY-MM-DD Lot): 2015-10-29   
        Vendor Name: FINISAR CORP.   
        Vendor OUI: 00-90-65
        Vendor PN: FTLX8571D3BCL   
        Vendor Rev: A   
        Vendor SN: AUJ0RCJ         


root@str-7260cx3-acs-7:/usr/local/lib/python3.9/dist-packages/sfputil# 

Help options
sfputil show eeprom-hexdump --help
root@str-7260cx3-acs-7:/usr/local/lib/python3.9/dist-packages/sfputil# sfputil show eeprom-hexdump --help
Usage: sfputil show eeprom-hexdump [OPTIONS]

  Display EEPROM hexdump of SFP transceiver(s) for a given port name and
  page number

Options:
  -p, --port <port_name>    Display SFP EEPROM hexdump for port <port_name>
                            [required]
  -n, --page <page_number>  Display SFP EEEPROM hexdump for
                            <page_number_in_hex>
  --help                    Show this message and exit.
root@str-7260cx3-acs-7:/usr/local/lib/python3.9/dist-packages/sfputil# 
```
Signed-off-by: Mihir Patel <patelmi@microsoft.com>

<!--
    Please make sure you've read and understood our contributing guidelines:
    https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

    ** Make sure all your commits include a signature generated with `git commit -s` **

    If this is a bug fix, make sure your description includes "closes #xxxx",
    "fixes #xxxx" or "resolves #xxxx" so that GitHub automatically closes the related
    issue when the PR is merged.

    If you are adding/modifying/removing any command or utility script, please also
    make sure to add/modify/remove any unit tests from the tests
    directory as appropriate.

    If you are modifying or removing an existing 'show', 'config' or 'sonic-clear'
    subcommand, or you are adding a new subcommand, please make sure you also
    update the Command Line Reference Guide (doc/Command-Reference.md) to reflect
    your changes.

    Please provide the following information:
-->

prgeor commented:
@andywongarista @keboliu please review

prgeor commented:
@mihirpat1 can you make page number optional? by default it should be page #0 so that non-paged optics also works fine

keboliu commented:
How to dump SFP/SFP+ cables(SFF-8472) EEPROM with this new command? Unlike QSFP/CMIS cables it has A0h and A2h address under page0.

prgeor commented:
@andywongarista could you please review and suggest how to make it work for 1G/10G SFP - SFF8472?

mihirpat1 commented:
> @mihirpat1 can you make page number optional? by default it should be page #0 so that non-paged optics also works fine

I have now addressed this with the new commit.

mihirpat1 commented:
> 1. How to dump SFP/SFP+ cables(SFF-8472) EEPROM with this new command? Unlike QSFP/CMIS cables it has A0h and A2h address under page0.

[MP] I have now modified the changes to display both A0h and A2h for SFF-8472 based transceivers. Thank you for catching this.

> 2. Could you please replace the pic in the description with formatted text?

[MP] I have addressed this now

andywongarista commented on sfputil/main.py:
```diff
@@ -688,6 +693,148 @@ def eeprom(port, dump_dom, namespace):
+            click.echo("Error: Failed to read EEPROM for A2h!")
+            sys.exit(ERROR_NOT_IMPLEMENTED)
+        else:
+            output += '\n\n{}A2h dump (Lower 128 bytes)'.format(indent)
```
Lowercase (for consistency with 'upper 128 bytes')?

mihirpat1 replied:
I have addressed this in the new commit now.

andywongarista commented on tests/sfputil_test.py:
```diff
@@ -418,6 +420,197 @@ def test_show_eeprom_RJ45(self, mock_chassis):
+        00000060 14 02 81 b1 0d 5a 17  a5 15 87 00 00 00 00 30 00 |.....Z........0.|
+        00000070 00 00 00 00 00 00 00  00 ff ff ff ff ff ff ff 01 |................|
+
+        A2h dump (upper 128 bytes) page 0h
```
From your testing it seems that page 2h has no output instead of something like this for page 0. Should we be consistent here for non-zero pages?

mihirpat1 replied:
The difference is due to the return value between page 0 v/s page 2 from the read_eeprom API. We can create a byte array for non-zero pages with all zeroes but that would also be more of modifying the result received from read_eeprom API. I think its better to leave the current output as is to keep the sanity of result received from read_eeprom API.

andywongarista commented on sfputil/main.py:
```diff
@@ -688,6 +693,148 @@ def eeprom(port, dump_dom, namespace):
+# 'eeprom-hexdump' subcommand
+@show.command()
+@click.option('-p', '--port', metavar='<port_name>', required=True, help="Display SFP EEPROM hexdump for port <port_name>")
+@click.option('-r', '--page', metavar='<page_number>', help="Display SFP EEEPROM hexdump for <page_number_in_hex>")
```
Why '-r'? This doesn't seem intuitive. Since '-p' is already taken for the port I would prefer '-n' for page **n**umber.

mihirpat1 replied:
I have addressed this in the new commit now.
